### PR TITLE
Adding a Fortran interface to libCEED

### PR DESCRIPTION
**Pull request by thilinarmtb**
Can you please review this @jedbrown @v-dobrev @tzanio ?

~I am trying to port C example `ex1` to Fortran. Once it is done I will add interfaces
for remaining functions and add tests for Fortran API as well~.

## TODO
- [x] Replace `malloc` and `realloc` by our own wrappers (which does error checking), Can we use
  `CeedMalloc` for malloc? (**Update**: I created a CeedRealloc to get this done)
- [ ] Conversion of Fortran strings to C strings. Currently, I do this by `fstring//char(0)` and that is ugly.
     ~Also I use `%val(0)' for `NULL`. This should be avoided as well.~
- [x] Add a Fortran header to define constants used in Ceed codes, e.g., `CEED_MEM_HOST`, etc.
- [x] Convert `CeedInt` and `CeedScalar` in Fortran interface to `int` and `double`.

### CODE REVIEW

**Comment by thilinarmtb** [comment]
There are a few bugs that needs to be fixed. I will work on them.

**Comment by thilinarmtb** [comment]
I guess travis is complaining because it can't find the Fortran compilers.

**Comment by thilinarmtb** [comment]
`t40-fortran` test is skipped locally, probably because the backends can't be used.
That's a bit strange.

**Comment by jedbrown** [comment]
It's possible to declare a Fortran integer that is the same size as a C address. That would be simpler so we don't have to have stubs translate them. Also, do we really want an untyped F77 interface, versus modules?

**Comment by thilinarmtb** [comment]
Thanks for the reply @jedbrown ! 

I like the idea of using a Fortran integer that is the same size as a C address
and will look into it.

We have never used modules with Nek5000 so far so I do not know enough
of it to comment. A quick google search gives me that they are a Fortran 90
feature. I am not sure how it will work with our Fortran77 code base.

I am trying to follow the Fortran interface of `gslib` here as well. I think it is
pretty user friendly and simple. I personally like it very much.

What do you think?

**Comment by jedbrown** [comment]
Does Nek actually use a strict F77 compiler or is it just F77 code compiled using a compiler that recognizes modern code?

PETSc recently surveyed this and found that all of our users were willing/eager to use F90 modules. We made the change in PETSc-3.8 (released this fall) and have had significantly fewer support requests due to Fortran users passing incorrect arguments. MPI now offers `mpi_f08` which also has better type safety. My inclination is to use a modern standard.

**Comment by thilinarmtb** [comment]
I have compiled Nek with newer versions of Fortran compilers other than F77.

I am not sure at this point what the complications for Nek would be if we start
using modules. I will get back to you on that after talking with some of the Nek
experts.

Also, is it possible to pass everything by reference in the function pointer passed
into `CeedQFunctionCreateInterior`? i.e., instead of,
```
int (*f)(void *ctx, void *qdata, CeedInt nq, const CeedScalar *const *u, CeedScalar *const *v)
```
in `CeedQFunctionCreateInterior`, can we have `nq` passed by reference as well?
```
int (*f)(void *ctx, void *qdata, CeedInt* nq, const CeedScalar *const *u, CeedScalar *const *v)
```
This will simplify most of the things when passing Fortran subroutines.

**Comment by jedbrown** [comment]
Instead of making the C interface confusing, I'd rather hold a separate Fortran function pointer.

    static int CeedQFunctionFortranStub(void *ctx, void *qdata, CeedInt nq, const CeedScalar *const *u, CeedScalar *const *v) {
      struct {
        void (*f)(void *ctx, void *qdata, CeedInt* nq, const CeedScalar *const *u, CeedScalar *const *v, int *ierr),
        void *innerctx;
      } *fctx = ctx;
      int ierr;
      fctx->f(fctx->innerctx, qdata, &nq, u, v, &ierr);
      return ierr;
    }

**Comment by jedbrown** [comment]
Has there been any progress on this? What would be the best way for me to help?

**Comment by thilinarmtb** [comment]
Hi Jed, I am still not sure about using the modules so I am discussing it with Paul.

I can continue working on a strict F77 interface but If we decide to use modules,
that will be a wastel. Hopefully we can clarify this issue pretty soon.

**Comment by jedbrown** [comment]
Yes, let's talk about it on the telecon tomorrow.

**Comment by jedbrown** [comment]
@thilinarmtb Can we move the Fortran stubs out of `ceed.h` and `ceed.c` into a new file like `ceed-fortran.c`? Also, do any of the subroutine names need to be in a header?

Should we have a header for Fortran users to include that provides types (`CeedScalar`, `CeedMemType`) or at least named constants like `CEED_MEM_DEVICE` and `CEED_TRANSPOSE`? Usage would be similar to `mpif.h`. Right now there are a lot of magic constants in user code.

**Comment by thilinarmtb** [comment]
@jedbrown Yes, I will move the Fortran stubs out of `ceed.c/.h`. 
No, the Fortran subroutine names need not be in a header.

I will add a header with named constants as well. I was going to do it after
porting `ex1`.

File: ceed.c [hunk 1]
```diff
@@ -217,3 +217,157 @@ void CeedDebug(const char *format,...) {
...
+static Ceed **Ceed_dict = NULL;
+static int Ceed_count = 0;
+static int Ceed_count_max = 0;
+
+void fCeedInit(const char* resource, CeedInt *ceed, CeedInt *err) {
+  if (Ceed_count == Ceed_count_max)
+    Ceed_count_max += Ceed_count_max/2 + 1,
+    Ceed_dict = realloc(Ceed_dict, sizeof(Ceed*)*Ceed_count_max);
+
+  Ceed *ceed_ = Ceed_dict[Ceed_count] = (Ceed*) malloc(sizeof(Ceed));
```
[the remark this answers is not in the document]
Comment (jedbrown) [reply ceed.c]: Also, are you rejecting the idea of matching the Fortran types?

```
#if CEED_SIZEOF_ADDR == 8
#  define CeedFortranAddr integer8
#else
#  define CeedFortranAddr integer4
#endif
```
(Possibly there is a way to do this from pure Fortran.)

File: name.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#ifndef NAME_H
```
[the remark this answers is not in the document]
Comment (thilinarmtb) [reply name.h]: Will do.

File: ceed.c [hunk 3]
```diff
@@ -217,3 +217,157 @@ void CeedDebug(const char *format,...) {
...
+    CeedQFunction_dict =
+        realloc(CeedQFunction_dict, sizeof(CeedQFunction*)*CeedQFunction_count_max);
+
+  CeedQFunction *qf_ = CeedQFunction_dict[CeedQFunction_count] =
+                          (CeedQFunction*) malloc(sizeof(CeedQFunction));
+
+  *err = CeedQFunctionCreateInterior(*Ceed_dict[*ceed], *vlength, *nfields,
+             *qdatasize, *inmode, *outmode, CeedQFunctionFortranStub,focca, qf_);
+
+  struct fContext fctx = {f, NULL};
```
[the remark this answers is not in the document]
Comment (thilinarmtb) [reply ceed.c]: Derp, I missed it. Will fix it.

**Comment by thilinarmtb** [comment]
@jedbrown: Thanks for the reviews. I will fix the issues.

File: ceed.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (thilinarmtb) [reply ceed.c]: I will remove the redundant indirection and will see if this is possible from pure Fortran.

File: ceed.c [hunk 4]
```diff
@@ -217,3 +217,157 @@ void CeedDebug(const char *format,...) {
...
+
+  Ceed *ceed_ = Ceed_dict[Ceed_count] = (Ceed*) malloc(sizeof(Ceed));
+  *err = CeedInit(resource, ceed_);
+
+  *ceed = Ceed_count++;
+}
+
+#define fCeedDestroy FORTRAN_NAME(ceeddestroy,CEEDDESTROY)
+void fCeedDestroy(CeedInt *ceed, CeedInt *err) {
+  *err = CeedDestroy(Ceed_dict[*ceed]);
```
[the remark this answers is not in the document]
Comment (thilinarmtb) [reply ceed.c]: Will fix it.

File: ceed-fortran.c [hunk 5]
```diff
@@ -69,14 +70,22 @@ void fCeedVectorSetArray(CeedInt *vec, CeedInt *memtype, CeedInt *copymode,
...
     FORTRAN_NAME(ceedvectorgetarrayread,CEEDVECTORGETARRAYREAD)
-void fCeedVectorGetArrayRead(CeedInt *vec, CeedInt *memtype, const CeedScalar *array,
+void fCeedVectorGetArrayRead(CeedInt *vec, CeedInt *memtype, CeedScalar *array,
     CeedInt *err) {
-  *err = CeedVectorGetArrayRead(CeedVector_dict[*vec], *memtype, &array);
+  const CeedScalar *b;
+  CeedVector vec_ = CeedVector_dict[*vec];
+  *err = CeedVectorGetArrayRead(CeedVector_dict[*vec], *memtype, &b);
+  if (*err == 0)
+    memcpy(array, b, sizeof(CeedScalar)*vec_->length);
```
[the remark this answers is not in the document]
Comment (jedbrown) [reply ceed-fortran.c]: The F77 solution isn't beautiful.

https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecGetArray.html

cf.
https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecGetArrayF90.html#VecGetArrayF90

[the remark this answers is not in the document]
Comment (thilinarmtb) [reply ceed-fortran.c]: Thanks for the links @jed. I will look into it.

**Comment by jed** [comment]
No worries, happy to help!

File: ceed-fortran.c [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+
+static int CeedQFunctionFortranStub(void *ctx, void *qdata, CeedInt nq,
+    const CeedScalar *const *u, CeedScalar *const *v) {
+  struct fContext *fctx = ctx;
+  int ierr;
+  //Note: Since u and v are arrays to pointers, they
+  //can't be passed as is to Fortran. So, what I do is
+  //find the first non-NULL pointer and pass it is
+  //address to Fortran function. On the Fortran side, to access
+  //specific type of data, the use has to calculate the indices.
```
Comment (jedbrown) [inline ceed-fortran.c]: I think the extensible way is to pass a base pointer and an array of offsets. Then you would write `u(offset(f) + i)` to correspond to `u[f][i]` (or `u[f-1][i-1]` if the Fortran indexing starts at 1).

An alternative is to just enumerate calls with explicit arguments, so the Fortran would write something like

    subroutine qf(ctx, qdata, nq, u1, u2, u3, v1, v2, ierr)

and the stub would have to invoke the correct number of arguments (not a big deal to enumerate up to 10 total arguments).

File: ceed-fortran.c [hunk 7]
```diff
@@ -64,41 +64,41 @@ void fCeedVectorCreate(int *ceed, int *length, int *vec, int *err) {
...
-    CeedScalar *array, int *err) {
+    double *array, int *err) {
   *err = CeedVectorSetArray(CeedVector_dict[*vec], *memtype, *copymode, array);
 }
 
 #define fCeedVectorGetArray FORTRAN_NAME(ceedvectorgetarray,CEEDVECTORGETARRAY)
 //TODO Need Fixing, double pointer
-void fCeedVectorGetArray(int *vec, int *memtype, CeedScalar *array, int *err) {
-  CeedScalar *b;
+void fCeedVectorGetArray(int *vec, int *memtype, double *array, int *err) {
```
Comment (jedbrown) [inline ceed-fortran.c]: Why this change? We'll have to revisit every such interface whenever we enable changing the `CeedScalar` typedef.

File: examples/nek5000/README.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+You can run Nek5000 examples by invoking `make-nek-examples.sh` script.
+Syntax for the command is 
+```sh
+  ./run-nek-example <example_name> <#mpi_ranks> <rea_name> <rea_and_map_path>"
+```
+
+Example:
+```
+  echo "Example ./run-nek-example ex1 4 b10 ./boxes/b10"
```
Comment (jedbrown) [inline examples/nek5000/README.md]: Let's make the example something a user can run with the boxes generated above. Also perhaps name the Nek5000 tools that need to be built (`genbox`, `genmap`, and `reatore2`).

I tried running `./run-nek-example.sh ex1 2 b2 boxes/b2` and get 

```
[...]
 set initial conditions
 nekuic (1) for ifld            1
 call nekuic for vel
 xyz min     0.0000       0.0000       0.0000
 uvwpt min   0.0000       0.0000       0.0000       0.0000       0.0000
 PS min     0.99000E+22
 xyz max     1.0000       1.0000       1.0000
 uvwpt max   0.0000       0.0000       0.0000       0.0000       0.0000
 PS max    -0.99000E+22
 done :: set initial conditions

 call userchk
GFM max:  3.5084E-02  3.5084E-02  8.7711E-03  2.5577E-16  8.2593E-17  8.2593E-17
  8        4  3.4780E-12  3.4780E-12  3.4780E-12  error a        9
  8        4  3.4780E-12  3.4780E-12  3.4780E-12  error b        9
  8        4  3.4780E-12  3.4780E-12  3.4780E-12  error c        9
case scalar:      2  7      2         4          1372       686   9  4.03041E-03  3.40412E+05  4.47823E-04  6.52803E-07
EXIT: quit after cggos          7

an error occured: dying ...


total elapsed time             :   3.68236E-02 sec
total solver time w/o IO       :   0.00000E+00 sec
time/timestep                  :   0.00000E+00 sec
avg throughput per timestep    :   0.00000E+00 gridpts/CPUs
total max memory usage         :   1.20783E-01 GB

-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
```

Do you know why?

**Comment by jedbrown** [comment]
I have this change after building the Nek example. This is bound to be committed accidentally. How can we make it so that files tracked by Git are not modified when building/running an example?

```
diff --git i/examples/nek5000/SIZE w/examples/nek5000/SIZE
index e573120..e7e6533 100644
--- i/examples/nek5000/SIZE
+++ w/examples/nek5000/SIZE
@@ -143,3 +143,7 @@ c automatically added by makenek
 c automatically added by makenek
       integer lcvx1,lcvy1,lcvz1,lcvelt
       parameter (lcvx1=1,lcvy1=1,lcvz1=1,lcvelt=1) ! cvode arrays
+
+c automatically added by makenek
+      integer ldimt_proj
+      parameter(ldimt_proj = 1) ! max auxiliary fields residual projection 
```

File: examples/nek5000/README.md (hunk 8 above)
Comment (thilinarmtb) [reply examples/nek5000/README.md]: Once we get our timing for BP1,we call `exitti` subroutine which exit with a non-zero value. That is why the error. It is not due to an error in the program itself. I can remove this call as it is confusing.

**Comment by jedbrown** [comment]
Thanks.  The example should exit successfully IMO.

File: examples/nek5000/README.md (hunk 8 above)
Comment (thilinarmtb) [reply examples/nek5000/README.md]: I will update the `run-nek-example` documentation as well.

File: ceed-fortran.c (hunk 7 above)
Comment (thilinarmtb) [reply ceed-fortran.c]: Derp, this was a mistake. I undid it.

**Comment by thilinarmtb** [comment]
Thank you very much for the review @jedbrown !!

The change in SIZE file is something added automatically by `makenek` and I am not sure
how to get rid of this. It does not affect subsequent builds though.

I undid the CeedScalar --> double conversion.

Let me know if this looks okay.

**Comment by jedbrown** [comment]
I pushed some minor changes; looks good to me now.

**Comment by thilinarmtb** [comment]
Thank you very much @jedbrown !! Looks good to me as well. +1 to merge.

**Comment by tzanio** [comment]
Great work guys, thanks @thilinarmtb and @jedbrown!
